File: app/src/main/java/com/kchen52/yetanothertranslinkapp/MapsActivity.java
package com.kchen52.yetanothertranslinkapp;

import android.Manifest;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Bundle;
import android.os.Message;
import android.preference.PreferenceManager;
import androidx.annotation.NonNull;
import com.google.android.material.snackbar.Snackbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.maps.android.data.kml.KmlLayer;
import com.kchen52.yetanothertranslinkapp.handlers.BusHandler;
import com.kchen52.yetanothertranslinkapp.handlers.HandlerExtension;
import com.kchen52.yetanothertranslinkapp.handlers.RequestHandler;

import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedList;

public class MapsActivity extends AppCompatActivity implements OnMapReadyCallback,
        GoogleMap.OnMyLocationClickListener, GoogleMap.OnMyLocationButtonClickListener {
    private final String TAG = "MapsActivity";
    private HandlerExtension handlerExtension;

    private GoogleMap mMap;
    private int MY_LOCATION_REQUEST_CODE = 1;

    // Used for formatting time/date for display
    private DateFormat dateFormat = new SimpleDateFormat("HH:mm:ss MM/dd/yyyy");

    private ArrayList<Marker> markers = new ArrayList<>();

    private RequestHandler requestHandler;

    private SharedPreferences sharedPref;

    @Override
    protected void onPause() {
        if (mMap != null) {
            CameraPosition cameraPosition = mMap.getCameraPosition();
            LatLng latLng = cameraPosition.target;
            SharedPreferences.Editor sharedPrefEditor = sharedPref.edit();
            sharedPrefEditor.putFloat("lastLat", (float) latLng.latitude);
            sharedPrefEditor.putFloat("lastLong", (float) latLng.longitude);
            sharedPrefEditor.putFloat("lastZoom", (float) cameraPosition.zoom);
            sharedPrefEditor.commit();
        }
        super.onPause();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_maps);

        handlerExtension = new HandlerExtension(this);

        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);


        sharedPref = PreferenceManager.getDefaultSharedPreferences(this);
        requestHandler = new RequestHandler(getApplicationContext());
    }

    @Override
    protected void onStart() {
        super.onStart();
        // Note: calling this code in onStart so changes made to the sharedpreferences in the settings or bus list screen are
        // immediately applied once coming back to the maps activity. onStart is also called after onCreate, so this also covers
        // that case lol
        requestHandler.update();
        // When the user chooses a new route in the bus list menu and returns to the main activity, draw that route
        if (mMap != null) {
            createRouteOverlays(requestHandler.getBusesRequested());
            centerCamera(
                    (double) sharedPref.getFloat("lastLat", 49.264566F),
                    (double) sharedPref.getFloat("lastLong", -123.133253F),
                    sharedPref.getFloat("lastZoom", 10f));
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.maps_activity_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        CameraPosition cameraPosition = mMap.getCameraPosition();
        LatLng latLng = cameraPosition.target;
        SharedPreferences.Editor sharedPrefEditor = sharedPref.edit();
        sharedPrefEditor.putFloat("lastLat", (float) latLng.latitude);
        sharedPrefEditor.putFloat("lastLong", (float) latLng.longitude);
        sharedPrefEditor.putFloat("lastZoom", cameraPosition.zoom);
        sharedPrefEditor.commit();
        switch (item.getItemId()) {
            case R.id.bus_list:
                Intent busListIntent = new Intent(this, BusListActivity.class);
                startActivity(busListIntent);
                return true;
            case R.id.settings:
                Intent settingsIntent = new Intent(this, SettingsActivity.class);
                startActivity(settingsIntent);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;

        centerCamera(
                (double) sharedPref.getFloat("lastLat", 49.264566F),
                (double) sharedPref.getFloat("lastLong", -123.133253F),
                sharedPref.getFloat("lastZoom", 10f));

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED) {
            mMap.setMyLocationEnabled(true);
        } else {
            // Show rationale and request permission.
            ActivityCompat.requestPermissions(this, new String[]{
                    Manifest.permission.ACCESS_FINE_LOCATION
            }, MY_LOCATION_REQUEST_CODE);
        }
        createRouteOverlays(requestHandler.getBusesRequested());
    }

    private void createRouteOverlays(String busesRequested) {
        if ("".equals(busesRequested)) {
            return;
        }
        // Clear the entire map
        mMap.clear();
        // Then create new layers and draw them
        for (String bus : busesRequested.split(", ")) {
            KmlLayer kmlLayer;
            try {
                int id = getResources().getIdentifier("raw/_" + bus.toLowerCase(), null, this.getPackageName());
                kmlLayer = new KmlLayer(mMap, id, getApplicationContext());
                kmlLayer.addLayerToMap();
            } catch (XmlPullParserException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }

        }

    }

    private void updateDisplayedTime(Date currentTime) {
        String formattedTime;
        if (currentTime == null) {
            formattedTime = "Last updated: Never";
        } else {
            String lastRequestedTime = dateFormat.format(currentTime);
            formattedTime = String.format(getResources().getString(R.string.last_updated_preamble), lastRequestedTime);
        }
        TextView lastRequestedTime_TextView = (TextView) findViewById(R.id.lastRequestDateTextView);
        lastRequestedTime_TextView.setText(formattedTime);
    }

    private void drawBuses(LinkedList<Bus> buses) {
        // If there are new buses, wipe the old ones from the map
        if (buses.size() != 0) {
            for (Marker marker : markers) {
                marker.remove();
            }
            markers.clear();
        }
        for (Bus bus : buses) {
            Log.i("Drawing the following:", "Destination: " + bus.getDestination() + ", VehicleNo: " + bus.getVehicleNumber() +
                    ", Longitude: " + bus.getLongitude() + ", Latitude: " + bus.getLatitude());

            addMarker(bus);
        }
    }

    private void centerCamera(double avgLat, double avgLong, float zoomLevel) {
        LatLng newLocation = new LatLng(avgLat, avgLong);
        mMap.moveCamera(CameraUpdateFactory.newLatLng(newLocation));
        mMap.animateCamera(CameraUpdateFactory.newLatLngZoom(newLocation, zoomLevel));
    }

    private void addMarker(Bus bus) {
        LatLng busLocation = new LatLng(bus.getLatitude(), bus.getLongitude());
        MarkerOptions temp = new MarkerOptions().position(busLocation)
                .title(bus.getDestination()+  ": " + bus.getVehicleNumber());

        if ("west".equals(bus.getDirection().toLowerCase())) {
            temp.icon(BitmapDescriptorFactory.fromResource(R.mipmap.ic_directions_bus_black_24dp_west));
        } else if ("east".equals(bus.getDirection().toLowerCase())) {
            temp.icon(BitmapDescriptorFactory.fromResource(R.mipmap.ic_directions_bus_black_24dp_east));
        } else {
            // And in the case it's a direction I can't parse, just show the default bus icon
            temp.icon(BitmapDescriptorFactory.fromResource(R.mipmap.ic_directions_bus_black_24dp));

        }
        markers.add(mMap.addMarker(temp));
    }

    /*
     * Displays a snackbar indicated what was done, then calls the action in RequestHandler
     */
    public void requestInformation(View view) {
        String busesRequested = requestHandler.getBusesRequested();
        if (busesRequested.equals("")) {
            Snackbar.make(view, "Please select at least one bus route to track.", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show();
            return;
        }

        if (!requestHandler.hasActiveInternetConnection()) {
            Snackbar.make(view, "This device currently doesn't have an internet connection. " +
                    "Please try again later.", Snackbar.LENGTH_LONG)
                    .setAction("Action", null).show();
            return;
        }
        requestHandler.updateWithInternet(handlerExtension);
    }

    // Takes the message from the Handler and updates the UI
    public void updateDisplayedTimeAndDrawBuses(Message message) {
        Bundle bundle = message.getData();
        BusHandler busHandler = bundle.getParcelable("busHandler");
        updateDisplayedTime(busHandler.getLastUpdatedTime());
        drawBuses(busHandler.getBuses());
    }

    @Override
    public boolean onMyLocationButtonClick() {
        return false;
    }

    @Override
    public void onMyLocationClick(@NonNull Location location) {
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions,
                                           int[] grantResults) throws SecurityException {
        if (requestCode == MY_LOCATION_REQUEST_CODE) {
            if (permissions.length == 1 &&
                    permissions[0] == Manifest.permission.ACCESS_FINE_LOCATION &&
                    grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                mMap.setMyLocationEnabled(true);
            } else {
                // Permission was denied. Display an error message.
            }
        }
    }
}

File: app/src/main/java/com/kchen52/yetanothertranslinkapp/map/MapsActivity.kt
package com.kchen52.yetanothertranslinkapp.map

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.kchen52.yetanothertranslinkapp.buslist.BusListActivity
import com.kchen52.yetanothertranslinkapp.R
import com.kchen52.yetanothertranslinkapp.SettingsActivity
import kotlinx.android.synthetic.main.activity_maps.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.collect

@ExperimentalCoroutinesApi
class MapsActivity : AppCompatActivity(), OnMapReadyCallback, MapsActivityView {
    private var googleMap: GoogleMap? = null

    private val coroutineScope = CoroutineScope(Dispatchers.Default + Job())
    private val viewModel: MapsActivityViewModel by viewModels { MapsActivityViewModelFactory(applicationContext, coroutineScope) }

    private lateinit var sharedPref: SharedPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_maps)

        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        val mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)
        sharedPref = getSharedPreferences(MapConstants.SHARED_PREFS_NAME, Context.MODE_PRIVATE)


        coroutineScope.launch {
            viewModel.getMapsActivityState().collect { mapsActivityState ->
                withContext(Dispatchers.Main) {
                    render(mapsActivityState)
                }
            }
        }

        refreshBusesButton.setOnClickListener {
            // Read buses from shared prefs
            val requestedBuses = sharedPref.getString("requestedBuses", "")?.map { it.toInt() }?.toIntArray() ?: intArrayOf()
            viewModel.onIntent(MapsActivityIntents.LoadBuses(requestedBuses))
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.maps_activity_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.bus_list -> {
                val busListIntent = Intent(this, BusListActivity::class.java)
                startActivity(busListIntent)
                true
            }
            R.id.settings -> {
                val settingsIntent = Intent(this, SettingsActivity::class.java)
                startActivity(settingsIntent)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        this.googleMap = googleMap
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
            == PackageManager.PERMISSION_GRANTED) {
            this.googleMap?.isMyLocationEnabled = true
        } else {
            // Show rationale and request permission.
            ActivityCompat.requestPermissions(this, arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION
            ), MY_LOCATION_REQUEST_CODE)
        }

        centerCamera(
            lat = sharedPref.getFloat("lastLat", MapConstants.DEFAULT_LATITUDE_VALUE),
            long = sharedPref.getFloat("lastLong", MapConstants.DEFAULT_LONGITUDE_VALUE),
            zoomLevel = sharedPref.getFloat("lastZoom", MapConstants.DEFAULT_ZOOM_LEVEL)
        )
    }


    @Throws(SecurityException::class)
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        if (requestCode == MY_LOCATION_REQUEST_CODE) {
            if (permissions.size == 1 && permissions[0] === Manifest.permission.ACCESS_FINE_LOCATION && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                googleMap?.isMyLocationEnabled = true
            } else {
                // TODO: Determine what should actually be done here. Display an error message?
            }
        }
    }


    override fun render(state: MapsActivityState) {
        runOnUiThread {
            when (state) {
                is MapsActivityState.DataState -> {
                    renderDataState(state)
                }
                is MapsActivityState.LoadingState -> {
                    renderLoading()
                }
                is MapsActivityState.ErrorState -> {
                    renderError(state.exception)
                }
            }
        }
    }

    private fun renderDataState(dataState: MapsActivityState.DataState) {
        // Unfortunately there's nothing we can do here :|
        // TODO: Can we await the map being ready?
        if (googleMap == null) {
            return
        }

        // Remove all existing markers first
        googleMap?.clear()

        // Only perform a camera shift if we have nonnull values for long, lat, and zoom
        if (dataState.latitude != null && dataState.longitude != null && dataState.zoom != null) {
            centerCamera(dataState.latitude, dataState.longitude, dataState.zoom)
        }

        loadingBusesProgressBar.visibility = View.GONE
        for (bus in dataState.buses) {
            googleMap?.addMarker(
                MarkerOptions()
                    .position(LatLng(bus.latitude, bus.longitude))
                    .title("${bus.destination} : ${bus.vehicleNo}")
                    .icon(BitmapDescriptorFactory.fromResource(R.mipmap.ic_directions_bus_black_24dp))
            )
        }
        val lastRequestTime = "Last requested: ${dataState.timeUpdated}"
        lastRequestDateTextView.text = lastRequestTime
    }

    private fun renderLoading() {
        loadingBusesProgressBar.visibility = View.VISIBLE
    }

    private fun renderError(exception: Exception) {
        loadingBusesProgressBar.visibility = View.GONE
        Toast.makeText(this, exception.message, Toast.LENGTH_LONG).show()
    }

    // Centres the Google Map object on the provided latitude, longitude, and at the zoom level
    // provided.
    private fun centerCamera(lat: Float, long: Float, zoomLevel: Float) {
        googleMap?.run {
            val newLocation = LatLng(lat.toDouble(), long.toDouble())
            moveCamera(CameraUpdateFactory.newLatLng(newLocation))
            animateCamera(CameraUpdateFactory.newLatLngZoom(newLocation, zoomLevel))
        }
    }

    // Save the current camera position so we go back to it the next time we resume
    override fun onPause() {
        super.onPause()
        googleMap?.let { googleMap ->
            val cameraPosition = googleMap.cameraPosition
            val latLng = cameraPosition.target
            val sharedPrefEditor = sharedPref.edit()
            sharedPrefEditor.putFloat("lastLat", latLng.latitude.toFloat())
            sharedPrefEditor.putFloat("lastLong", latLng.longitude.toFloat())
            sharedPrefEditor.putFloat("lastZoom", cameraPosition.zoom)
            sharedPrefEditor.apply()
        }
    }


    companion object {
        private const val TAG = "MapsActivity"
        private const val MY_LOCATION_REQUEST_CODE = 1
    }
}